src/Traverser.php

```diff
@@ -76,13 +76,13 @@ public function children(): Collection
         return $this->resolveRelation('children', collect())->filter();
     }
 
-    protected function inferChildren($objects): Collection
+    public function inferChildren($objects): Collection
     {
-        return collect($objects)
-            ->filter(function ($object) {
-                return $parent = static::make($object, $this->relations)->parent()
-                    && $this->is($parent);
-            });
+        return collect($objects)->filter(function ($object) {
+            $parent = static::make($object, $this->relations)->parent();
```

Comment:
This is causing a recursive loop when $objects contains Objects with inferred parents.

Reply:
Solved this by being more specific about what is passed into the inferParent and inferChildren methods.

- inferParent must only be passed objects that can actually be a parent
- inferChildren must only be passed objects that can actually be children